src/typecheck.c

```diff
@@ -815,6 +820,62 @@ static bool enum_member_exists(const Type *t, const char *name)
     return false;
 }
 
+static const Type *cast_array_members_to_a_common_type(Location error_location, ExpressionTypes **exprtypes)
+{
+    // Avoid O(ntypes^2) code in a long array where all or almost all items have the same type.
+    // This is at most O(ntypes*k) where k is the number of distinct types.
+    List(const Type *) distinct = {0};
+    for (ExpressionTypes **et = exprtypes; *et; et++) {
+        bool found = false;
+        for (const Type **t = distinct.ptr; t < End(distinct); t++) {
+            if ((*et)->type == *t) {
+                found = true;
+                break;
+            }
+        }
+        if (!found)
+            Append(&distinct, (*et)->type);
+    }
+
+    const Type *result = NULL;
+    for (const Type **t = distinct.ptr; t < End(distinct); t++) {
+        bool t_compatible_with_all_others = true;
+        for (const Type **t2 = distinct.ptr; t2 < End(distinct); t2++) {
+            if (!can_cast_implicitly(*t2,*t)) {
+                t_compatible_with_all_others = false;
+                break;
+            }
+        }
+        if (t_compatible_with_all_others) {
+            /*
+            This assertion should never fail, assuming that
+
+                can_cast_implicitly(A,B) && can_cast_implicitly(B,A)
+
+            is possible only when A and B are the same type.
```

Comment:
This assumption isn't always true, consider `int*` and `void*`. A list containing both of these should probably have item type `void*`, since it would turn into a `void*` list when a new item is added.